### PR TITLE
sklearn_tools set added.

### PR DESCRIPTION
**Pull request by pkohvaei**

### CODE REVIEW

File: tools/sklearn_tools/discriminant.xml [hunk 1]
```diff
@@ -0,0 +1,199 @@
+<tool id="discriminant_classifier" name="Discriminant Analysis" version="@VERSION@">
+    <description></description>
+    <macros>
+        <token name="@VERSION@">0.9</token>
```
Comment (bgruening) [inline tools/sklearn_tools/discriminant.xml]: Can we move the version macro into the main macro file? So the entire suite gets a main version number?

File: tools/sklearn_tools/discriminant.xml [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+        <macro name="tol" token_default_value="0.00001">
+            <param argument="tol" type="float" optional="true" value="@DEFAULT_VALUE@" label="Tollerance"
+                help="Rank estimation threshold used in SVD solver. ( default : @DEFAULT_VALUE@ )"/>
+        </macro>
+        <!--macro name="priors"-->
+    </macros>
+    <requirements>
```
Comment (bgruening) [inline tools/sklearn_tools/discriminant.xml]: move to main macro

File: tools/sklearn_tools/discriminant.xml [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+        <requirement type="package" version="2.3.0">anaconda</requirement>
+    </requirements>
+    <stdio>
```
Comment (bgruening) [inline tools/sklearn_tools/discriminant.xml]: move to main.macro

File: tools/sklearn_tools/discriminant.xml [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+        <exit_code level="fatal" range="1:"/>
+    </stdio>
+    <version_command>echo "@VERSION@"</version_command>
```
Comment (bgruening) [inline tools/sklearn_tools/discriminant.xml]: To main.macro?

File: tools/sklearn_tools/discriminant.xml [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+    <command><![CDATA[
+    export PYTHONPATH=/home/kohvaeip/.local/lib/python2.7/site-packages:/home/kohvaeip/.local/lib/:\$PYTHONPATH
+    &&
+    cat "$inputs" >&2
```
Comment (bgruening) [inline tools/sklearn_tools/discriminant.xml]: I added this just for debugging purpose, I guess we can remove this.

File: tools/sklearn_tools/discriminant.xml [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+            <output name="outfile" file="linear_prediction_result01.tabular"/>
+        </test>
+        <test>
+            <param name="infile_model" value="linear_model02.txt" ftype="txt"/>
+            <param name="infile_data" value="test_set.tabular" ftype="tabular"/>
+            <param name="selected_task" value="load"/>
+            <output name="outfile" file="linear_prediction_result02.tabular"/>
+        </test>
+        <!--test>
+            <param name="infile_model" value="linear_model03.txt" ftype="txt"/>
```
Comment (bgruening) [inline tools/sklearn_tools/discriminant.xml]: What was wrong with these tests? Can we enable them or remove them?

File: tools/sklearn_tools/discriminant.xml [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+        <test>
+            <param name="infile_model" value="quadratic_model01.txt" ftype="txt"/>
+            <param name="infile_data" value="test_set.tabular" ftype="tabular"/>
+            <param name="selected_task" value="load"/>
+            <output name="outfile" file="quad_prediction_result01.tabular"/>
+        </test>
+    </tests>
+    <help><![CDATA[
+**What it does**
+Linear Discriminant Analysis and Quadratic Discriminant Analysis are two classic classifiers, with, as their names suggest, a linear and a quadratic decision surface, respectively. This module uses sklearn.discriminant_analysis package to implement Discriminant Analysis classification.
```
Comment (bgruening) [inline tools/sklearn_tools/discriminant.xml]: as their names suggest, -> really ;)

Comment (bgruening) [inline tools/sklearn_tools/discriminant.xml]: A little bit information about the input and outputs would be nice. Format, scipy, can be used by ...

File: tools/sklearn_tools/discriminant.xml [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+
+
+    ]]></help>
+    <citations>
```
Comment (bgruening) [inline tools/sklearn_tools/discriminant.xml]: move to macros

File: tools/sklearn_tools/eden_converter.xml [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+    </requirements>
+    <stdio>
+        <exit_code level="fatal" range="1:"/>
+    </stdio>
+    <version_command>echo "@VERSION@"</version_command>
+    <command>
+        <![CDATA[
+        export PYTHONPATH=/home/kohvaeip/.local/lib/python2.7/site-packages:/home/kohvaeip/.local/lib/:\$PYTHONPATH
+        &&
+        cat "$inputs" >&2
```
Comment (bgruening) [inline tools/sklearn_tools/eden_converter.xml]: debugging ...

File: tools/sklearn_tools/eden_converter.xml [hunk 10]
```diff
@@ -0,0 +1,123 @@
+<tool id="eden_converter" name="Convert" version="@VERSION@">
+    <description>fasta sequences to graphs</description>
+    <macros>
+        <token name="@VERSION@">0.9</token>
+    </macros>
+    <requirements>
```
Comment (bgruening) [inline tools/sklearn_tools/eden_converter.xml]: moving stuff to macro file

File: tools/sklearn_tools/discriminant.xml [hunk 11]
```diff
@@ -0,0 +1,199 @@
+<tool id="discriminant_classifier" name="Discriminant Analysis" version="@VERSION@">
```
Comment (bgruening) [inline tools/sklearn_tools/discriminant.xml]: maybe change the id to sklearn_discriminant_classifier

File: tools/sklearn_tools/eden_converter.xml [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+    graph_list = rnafold_to_eden(sequence_iterable)
+elif "$converter_options.selected_converter" == "rnashapes_to_eden":
+    params["converter_options"].pop('selected_converter')
+    sub_params = params["converter_options"]["options"]
+    graph_list = rnashapes_to_eden(sequence_iterable, **sub_params )
+
+json_graph_list = eden_to_node_link_data(graph_list)
+
+#*TODO: replace with eden_to_node_link_file(graph_list, fname)*#
+with open("$outfile", 'w+') as out_file:
```
Comment (bgruening) [inline tools/sklearn_tools/eden_converter.xml]: I like when you are using `with` :)

File: tools/sklearn_tools/eden_converter.xml [hunk 13]
```diff
@@ -0,0 +1,123 @@
+<tool id="eden_converter" name="Convert" version="@VERSION@">
```
Comment (bgruening) [inline tools/sklearn_tools/eden_converter.xml]: This is the sequence specific converter, right? If you plan to have an additional one for molecules we need to change the id and make it more explicit.

File: tools/sklearn_tools/eden_converter.xml [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+    <description>fasta sequences to graphs</description>
+    <macros>
+        <token name="@VERSION@">0.9</token>
+    </macros>
+    <requirements>
+        <requirement type="package" version="2.3.0">anaconda</requirement>
```
Comment (bgruening) [inline tools/sklearn_tools/eden_converter.xml]: here rnafold and co is missing ...

File: tools/sklearn_tools/eden_vectorizer.xml [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+    </requirements>
+    <stdio>
+        <exit_code level="fatal" range="1:"/>
+    </stdio>
+    <version_command>echo "@VERSION@"</version_command>
+    <command>
+        <![CDATA[
+        export PYTHONPATH=/home/kohvaeip/.local/lib/python2.7/site-packages:/home/kohvaeip/.local/lib/:\$PYTHONPATH
+        &&
+        cat "$inputs" >&2
```
Comment (bgruening) [inline tools/sklearn_tools/eden_vectorizer.xml]: as above ...

File: tools/sklearn_tools/nn_classifier.xml [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+    </requirements>
+    <stdio>
+        <exit_code level="fatal" range="1:"/>
+    </stdio>
+
+    <version_command>echo "@VERSION@"</version_command>
+    <command><![CDATA[
+    export PYTHONPATH=/home/kohvaeip/.local/lib/python2.7/site-packages:/home/kohvaeip/.local/lib/:\$PYTHONPATH
+    &&
+    cat "$inputs" >&2
```
Comment (bgruening) [inline tools/sklearn_tools/nn_classifier.xml]: as above ...

File: tools/sklearn_tools/nn_classifier.xml [hunk 17]
```diff
@@ -0,0 +1,201 @@
...
+        </test>
+    </tests>
+    <help><![CDATA[
+**What it does**
+This module implements the k-nearest neighbors classification algorithms.
+For more information check http://scikit-learn.org/stable/modules/neighbors.html
+    ]]></help>
+    <citations>
+        <citation type="bibtex">
+            @article{scikit-learn,
```
Comment (bgruening) [inline tools/sklearn_tools/nn_classifier.xml]: macro

File: tools/sklearn_tools/numeric_clustering.xml [hunk 18]
```diff
@@ -0,0 +1,406 @@
+<tool id="numeric_clustering" name="Numeric Clustering" version="@VERSION@">
```
Comment (bgruening) [inline tools/sklearn_tools/numeric_clustering.xml]: sklearn prefix?

File: tools/sklearn_tools/numeric_clustering.xml [hunk 19]
```diff
@@ -0,0 +1,406 @@
...
+            <param name="end_column" value="1" />
+            <param name="n_clusters" value="4" />
+            <param name="random_state" value="100"/>
+            <output name="outfile" file="cluster_result16.txt"/>
+        </test>
+    </tests>
+    <help><![CDATA[
+**What it does**
+This tool offers different clustering algorithms which are provided by
+scikit-learn to find similarities among samples and cluster the samples based on these similarities.
```
Comment (bgruening) [inline tools/sklearn_tools/numeric_clustering.xml]: At some point we could include a guide of advantage and disadvantage about all these algorithms.
@pkohvaei this tool in particular just rocks!

**Comment by bgruening** [comment]
This PR is great. If you can fix the macros and ids I'm ready to merge this!

Thanks!
Bjoern

File: tools/sklearn_tools/discriminant.xml (hunk 11 above)
Comment (pkohvaei) [inline tools/sklearn_tools/discriminant.xml]: Changed!

File: tools/sklearn_tools/discriminant.xml (hunk 1 above)
Comment (pkohvaei) [inline tools/sklearn_tools/discriminant.xml]: Done for all tools.

File: tools/sklearn_tools/discriminant.xml (hunk 3 above)
Comment (pkohvaei) [inline tools/sklearn_tools/discriminant.xml]: For all tools.

File: tools/sklearn_tools/numeric_clustering.xml (hunk 18 above)
Comment (pkohvaei) [inline tools/sklearn_tools/numeric_clustering.xml]: There!

File: tools/sklearn_tools/discriminant.xml (hunk 5 above)
Comment (pkohvaei) [inline tools/sklearn_tools/discriminant.xml]: I removed it from all wrappers.

File: tools/sklearn_tools/nn_classifier.xml (hunk 17 above)
Comment (pkohvaei) [inline tools/sklearn_tools/nn_classifier.xml]: Done for all.

File: tools/sklearn_tools/eden_converter.xml (hunk 13 above)
Comment (pkohvaei) [inline tools/sklearn_tools/eden_converter.xml]: eden_sequence_converter

File: tools/sklearn_tools/discriminant.xml (hunk 4 above)
Comment (pkohvaei) [inline tools/sklearn_tools/discriminant.xml]: Sure!

File: tools/sklearn_tools/discriminant.xml (hunk 2 above)
Comment (pkohvaei) [inline tools/sklearn_tools/discriminant.xml]: Moved.

**Comment by bgruening** [comment]
We need to get the dependencies in and iterate over the filetypes at some point, but this is very nice start.

Thanks!
Bjoern
